File: backend/app/models/application.py
from datetime import datetime

from sqlalchemy import (
    DateTime,
    ForeignKey,
    Integer,
    String
)
from sqlalchemy.orm import (
    Mapped,
    mapped_column,
    relationship
)

from app.db.base import Base


class Application(Base):
    __tablename__ = "applications"

    id: Mapped[int] = mapped_column(
        Integer,
        primary_key=True,
        index=True
    )

    student_id: Mapped[int] = mapped_column(
        ForeignKey("student_profiles.user_id"),
        nullable=False
    )

    opportunity_id: Mapped[int] = mapped_column(
        ForeignKey("opportunities.id"),
        nullable=False
    )

    status: Mapped[str] = mapped_column(
        String(50),
        default="Applied",
        nullable=False
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=datetime.utcnow
    )

    student = relationship(
        "StudentProfile"
    )

    opportunity = relationship(
        "Opportunity"
    )

File: frontend/src/types/application.ts
import { Opportunity } from "./opportunity";
import { StudentProfile } from "./student-profile";

export interface Application {
  id: number;

  student: StudentProfile;

  opportunity: Opportunity;

  status: string;

  created_at: string;
}

export interface CreateApplicationRequest {
  opportunity_id: number;
}

export interface UpdateApplicationStatusRequest {
  status: string;
}